### ISSUE TITLE
Intl Support

### ISSUE THREAD

**Issue by danilobuerger** [facebook/hermes#23]
In https://github.com/facebook/hermes/blob/master/doc/Features.md `Intl` is listed as `Excluded From Support`. With 
> Hermes plans to target ECMAScript 2015 (ES6), with some carefully considered exceptions.

What are those considerations regarding `Intl`? (If its about size, why not have a variant like jsc does?)
How should i18n look like with Hermes?

**Comment by fbartho** [comment]
Intl is super important for us, and unfortunately, it's a blocker for my team. The Polyfills we tried before switching to jsc-intl weren't complete (they allowed compilation/basic testing, but didn't work in actual localized environments).

Is there a plan for when Intl might be discussed again for the Hermes Engine?

**Comment by dulinriley** [comment]
It will help us prioritize work on `Intl` if we know the following things:
* Its main use cases (most popular function it has) to the RN community
* Any important libraries that depend on it
* What alternatives exist (is there a JS-only way to implement it?)

I don't remember the exact conversation we had about `Intl`, but I think it was related to ICU's binary size, and that it is not part of the normal EcmaScript JS spec.

@avp or @ridiculousfish have more context on this

**Comment by danilobuerger** [comment]
> Its main use cases (most popular function it has) to the RN community

Enabling I18n for apps regarding:
* Date / Time Formats
* Relative Time Formats
* List Formats
* Number (also Currency) Formats
* Plural Rules

> Any important libraries that depend on it

https://github.com/formatjs/react-intl (> 10K Stars)
https://github.com/moment/luxon (> 8K Stars)

> What alternatives exist (is there a JS-only way to implement it?)

As pointed out by @fbartho, polyfilling doesn't work properly. Every other solution is incomplete at least regarding locale rules around the world.

> but I think it was related to ICU's binary size

If it is just about binary size, it could be done like JSC: a variant with Intl.
See https://github.com/react-native-community/jsc-android-buildscripts/#international-variant

> and that it is not part of the normal EcmaScript JS spec

However, lots of functions that are part of the ECMAScript 2015, are dependent on Intl to function "properly", or at least need a locale aware implementation anyway. For example:

> 20.3.4.39 Date.prototype.toLocaleString ( [ reserved1 [ , reserved2 ] ] )
An ECMAScript implementation that includes the ECMA-402 Internationalization API must implement the Date.prototype.toLocaleString method as specified in the ECMA-402 specification. If an ECMAScript implementation does not include the ECMA-402 API the following specification of the toLocaleString method is used.
> 
> This function returns a String value. The contents of the String are implementation-dependent, but are intended to represent the Date in the current time zone in a convenient, human-readable form that corresponds to the conventions of the host environment’s current locale.

**Comment by TheTimeWalker** [comment]
One thing to add: Intl works perfectly out of box on iOS as it has to use the V8 engine which already integrates this. So I find it strange why we're trying to back off from Intl support as this would just confuse RN developers with the iOS and Android feature gap.

**Comment by fbartho** [comment]
That's a great point @TheTimeWalker -- this is a "platform compatibility issue" where iOS & Android's JavaScript engines differ. (Nit: iOS uses the JavaScriptCore engine rather than the V8 engine, but your point is the same).

I guess the key question is: What is this project's driving mandate with regards to differences between iOS & Android? -- Presumably the expectation is that there shouldn't be significant framework-level differences? Especially not in things that can't be polyfilled-well.

Unfortunately, the existing Polyfills in this case are insufficient. React-Native developers targeting more than one language will be stuck on jsc-intl unless this feature is provided by Hermes!

**Comment by gmaclennan** [comment]
As @danilobuerger clearly states, there is no complete and reliable JS-only solution to this. For I18n apps `Intl` is needed. It adds to binary size, but as I understand it that's because I18n is so darn complicated with so many edge-cases.

**Comment by wyzzy** [comment]
Strong +1 for including `Intl` support out-of-the-box in Hermes, having just encountered this limitation ourselves!

At the moment the lack of `Intl` support on Android RN is a major source of inconsistency in behaviour between iOS and Android. On iOS `Intl` 'just works' as expected, but not so on Android, where you're stuck trying to find some workable solution to restore parity in behaviour on the two platforms. On iOS, `Number.prototype.toLocaleString()` for formatting numbers and currency values all work as expected, as does `Date.prototype.toLocaleString()` for dates, and all the `Intl` API...

The lack of `Intl` support on Android's out-of-the-box JSC was one significant hurdle we encountered in enabling our iOS-first RN app for Android. As many folks have said, the pollyfills just don't work well enough and feel like an unsatisfactory hack.

For our investment app, **date/time** and **currency formatting** are the primary drivers, though relative time rules, plural rules, and list formats would all undeniably help offer a more polished user experience.

Needing to include a custom JSC build on Android is non-trivial and really shouldn't be necessary for such a fundamental piece of functionality! 

If Hermes is worried about size, some RN config to optionally exclude `Intl` support for those (rare?) apps that don't need it might be a way forward. Surely these days most commercial quality apps need to reliably present localised time & date or numeric/currency values?

If Hermes did fully support `Intl`, switching from the existing JSC would be a no-brainer for us. On the other hand, no support means we would perhaps be better off with a JSC build that does support it?? Feels like a problem that Hermes needs an answer to, one way or the other.

I18n certainly is darn complicated, which is precisely why it's so important that the platform itself offers first-class support for `Intl` out of the box! Even on Android.

**Comment by learnyst** [comment]
Hi,
We also need the support of Intl in hermes.

**Comment by jcurtis** [comment]
Intl is a major concern for all the apps I have ever worked on. Today we're using [`react-intl`](https://github.com/formatjs/react-intl) which is one of the most popular React libraries for handling intl. As stated before, there are no polyfills that work properly on Android right now. 

I'd be curious if someone could tell us what Facebook uses to handle intl in their React Native apps? Specifically for number and date formatting.

**Comment by gpawlik** [comment]
@dulinriley is there any more information that you need in order to prioritise this issue?

**Comment by feruzm** [comment]
Big thanks to this thread, Hermes is great but we need support for `Intl` which most fundamental part of any app.

**Comment by OtacilioN** [comment]
@marlonchalegre

**Comment by OtacilioN** [comment]
Hey folks, 

Is that any workaround suggested for ```Number.prototype.toLocaleString()``` while they don't implement ```Intl``` ?  

We broke our minds for some weeks with this bug because when we were using the debug mode in React Native, the ```Number.prototype.toLocaleString()``` works, I think because it uses my browser engine to run JavaScript, but as soon we assemble the app or turnOff the debug mode, ```Number.prototype.toLocaleString()``` starts to no work.

**Comment by fbartho** [comment]
@OtacilioN -- On my team we're still using `org.webkit:android-jsc-intl:r245459` until Hermes supports `Intl` (we would love to use Hermes for Android Performance reasons!!) Not to give you useless advice, but that's what I recommend.

And I ran into exactly the same debug-time issues with some other features including Proxy (also missing in Hermes, #33), before that was updated to be available on android-jsc. I feel your pain!

**Comment by OtacilioN** [comment]
Yeah, we had some bugs related to proxy too, fortunately, checked the code of the lib that we were using, and made a pull request to not use proxy when is not supported, actually, the usage of proxy was not really necessary in the lib.

Unfortunately, we made the decision of migrating to Hermes a couple of months ago and getting back to JSC could be really painful now. I could not imagine that an engine created by facebook would not support localization. 😢

**Comment by mars-lan** [comment]
We ended up using [react-native-v8](https://github.com/Kudo/react-native-v8) which provides similar performance benefit like Hermes but has full intl support.

**Comment by patrickkempff** [comment]
How does facebook handle Intl in their apps when using Hermes?

**Comment by thinklinux** [comment]
@patrickkempff I don't think they use it in their apps.

**Comment by tmikov** [comment]
Sorry about the delayed response here. Internally at Facebook we use `fbt`, with its own specific translation assets handling, and translation dictionary shipped with the app binary. fbt was open sourced a while back, and fortunately we have RN android variant in OSS now. 

See:
https://github.com/facebookincubator/fbt
https://github.com/facebookincubator/fbt/tree/master/packages/fbt-rn-android-native
https://github.com/facebookincubator/fbt/tree/rn-demo-app

The published version is currently builds for Android but AFAIK there is nothing preventing porting to iOS. 

On the subject of Intl support: we would like to have it in Hermes, but it isn't our top priority, since, as described above, we don't use it internally. We welcome help from the community in implementing it.

**Comment by TheSavior** [comment]
We are revisiting `Intl` support on Hermes and need your help figuring out which functions to prioritize. There are a ton of `Intl` APIs and I'd bet that only a handful of them are the most critical functions blocking Hermes adoption.

Please help us understand the *particular* function calls that are *blocking* your ability to use Hermes. For example, I've seen `Number.prototype.toLocaleString()` mentioned twice in this thread. That's the kind of specific example I'm looking for. "Relative Time Formats" is too vague. I'm sure there are many more functions that would be really nice to have but aren't as critical. Please mention them, but separately so I can track that.

**Intl functions blocking usage of Hermes:**
- `Number.prototype.toLocaleString()`

**Intl functions really nice to have in Hermes:**


-----------------

Also, I saw a mention of the [react-intl](https://github.com/formatjs/react-intl) project. Can someone go through that and identify what `Intl` functions it uses and share those here?


Thanks!

**Comment by fbartho** [comment]
I'm not entirely sure how to search `react-intl` for "particular function calls" that need to be supported. I'm trying to join their slack to figure that out.

In addition to react-intl, a super popular library is MomentJS -- https://momentjs.com & its sub library moment-timezone https://momentjs.com/timezone/

Moment has a CI testing various browsers/environments, so maybe you can add hermes to that list, and then get an enumeration of the methods that are missing / fail tests?

**Comment by dlebedynskyi** [comment]
@TheSavior  
I believe we  are specifically interested in DateTime and TimeZone related ones such as 
* Date.prototype.toLocaleString
* Date.prototype.toLocaleTimeString
* Intl.DateTimeFormat().resolvedOptions
*  rest of timezone related things

For react-intl it is probalby better to start with [intl-messageformat](https://github.com/formatjs/formatjs/tree/master/packages/intl-messageformat) as it is most used one. 

@fbartho 
I believe moment is not actually using Intl.

**Comment by fbartho** [comment]
Our codebase also directly uses 

- [`String.prototype.localeCompare`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare)

**Comment by janicduplessis** [comment]
I'm currently using the following polyfills for Hermes:

- intl (https://github.com/andyearnshaw/Intl.js)
- @formatjs/intl-pluralrules (https://github.com/formatjs/formatjs/tree/master/packages/intl-pluralrules)
- @formatjs/intl-relativetimeformat (https://github.com/formatjs/formatjs/tree/master/packages/intl-relativetimeformat)

I'm using https://lingui.js.org/ as a higher level abstraction over intl apis.

I don't have anything blocking usage of hermes since it works fine with these polyfills. First thing would be to try and remove usage of the intl polyfill, this one actually implements a lot of apis, the list is available here https://github.com/andyearnshaw/Intl.js#implemented.

Here are some Intl apis I use either directly or through lingui-js:
- Intl.NumberFormat for currency formatting.
- Intl.DateTimeFormat for date formatting.
- Intl.RelativeTimeFormat for relative date formatting (x days ago).

**Comment by TheSavior** [comment]
@dlebedynskyi thanks, from my quick searching it looks like IntlMessageFormat isn't part of a finalized spec. Am I looking in the wrong place?

**Comment by TheSavior** [comment]
@janicduplessis, it looks like the Intl APIs you list are a collection of multiple functions. Do you know which functions in those APIs you use? The more specific the better here :)

**Comment by jozan** [comment]
I use [Luxon](https://moment.github.io/luxon/) and it uses Intl apis quite a bit. I assume lots of other people also use Luxon for handling dates and timezones and other things.

Luxon uses or supports at least these Intl methods:

- `Intl.RelativeTimeFormat` (constructor)
- `Intl.RelativeTimeFormat.prototype.format`
- `Intl.RelativeTimeFormat.prototype.formatToParts`
- `Intl.DateTimeFormat` (constructor)
- `Intl.DateTimeFormat.prototype.formatToParts`
- `Intl.DateTimeFormat.prototype.format`
- `Intl.DateTimeFormat.prototype.resolvedOptions`
- `Date.prototype.toLocaleString`

**Comment by janicduplessis** [comment]
@TheSavior Intl.XFormat apis actually don't really have many functions. AFAIK the usage is pretty much always:

```js
new Intl.XFormat('locale').format(thing)
```

For example `Intl.NumberFormat`:

```js
new Intl.NumberFormat('de-DE').format(number)
```

Actually I think what you are looking for is which options on the format constructor we want to support first. Here's a list of options that should be supported / I'm using

#### NumberFormat (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat/NumberFormat)

- style: I think all 4 options would be important to have
- currency: needed for currency support
- currencyDisplay
- minimumFractionDigits
- maximumFractionDigits

#### DateTimeFormat (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat/DateTimeFormat)

- dateStyle
- timeStyle
- hour12
- weekday, year, month, day, hour, minute, second

#### RelativeTimeFormat (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RelativeTimeFormat/RelativeTimeFormat)

This one doesn't have much options so I guess everything.

**Comment by TheSavior** [comment]
Thanks @janicduplessis @jozan, updated

**Comment by cdsanchez** [comment]
@TheSavior Would it be feasible for this `Intl` support to be built on top of JSI interfaces such that other JS engines with JSI implementations (beyond Hermes) can take advantage of this?

**Comment by TheSavior** [comment]
It looks like the [ChromeStatus page has usage numbers](https://www.chromestatus.com/metrics/feature/popularity) for these namepsaces: 

- DateTimeFormat 8.755547%
- Collator 5.05094%
- NumberFormat 3.182229%
- RelativeTimeFormat 0.165491%
- PluralRules 0.136803%
- ListFormat 0.00172%
- Locale <=0.000001%

It looks like there is a big drop off between NumberFormat and RelativeTimeFormat so that is probably a reasonable prioritization point.

**Comment by longlho** [comment]
Hello, I'm the maintainer of `react-intl` & `formatjs` & also a delegate on TC39/ECMA-402. We got quite a few requests to support non-traditional JS engines that don't have `Intl`, and sounds like `hermes` is 1 of those. There're no hierarchy of which Intl APIs are more important/popular from my perspective because most implementations either use `libicu` or has native bindings that wraps around `icu4j`/`icu4c`.

On top of that the standards are evolving. `formatjs` has polyfills for stage-3 APIs like Unified NumberFormat which allows you to format units like `5 feet` or compact notation like `1 thousand`. What you'd gain from native implementation in hermes vs polyfill would basically means hermes would have to package w/ CLDR data (& follow its upgrade cadence as well). That's a non-trivial amount of bytes. 

Performance-wise I don't see a huge advantage unless you're handling subsecond re-rendering.

The other reason why it's hard to pick and choose 2, 3 or 4 APIs to support is because if you look at the spec, they share a lot of abstract operations & CLDR data. For example, you typically always have to start with locale negotiation, which is the process to match what the user asks for, and what data you have. None of the browsers package ALL of CLDR so this is especially important. From there you start figuring out CLDR dataset(s) based on the API and then join them together to produce the `best-effort` output (that's right, ECMA-402 is best effort & not correct).

If I were to rank, the sequence you basically need would be something like:
1. `Intl.Locale` because you need locale negotiation, whether public or abstract
2. `Intl.DateTimeFormat` because you need 2 datasets: CLDR for pattern & IANA for timezone so it might help a lot of people to not have to polyfill those giant datasets.
3. `Intl.PluralRules` because it's very foundational, even to `NumberFormat` (to determine things like `1 dollar` vs `2 dollars`)
4. `Intl.NumberFormat` because it's an important primitive
5. `Intl.RelativeTimeFormat` is cherry-on-top.

The class of problem involving string like `Collation` algorithm is fairly separate and not necessarily required.

Happy to chat more :)

**Comment by janicduplessis** [comment]
@TheSavior I think another thing to consider is the binary size impact of each Intl api.

**Comment by longlho** [comment]
Not to mention CLDR data compression in and of itself is a whole problem set, because of how locale fallback works (e.g `en-GB` -> `en-001` -> `en`) so yeah, the increase in size won't be trivial.

We're having debate in ECMA-402 during Unified NumberFormat on *what* unit data to package, which is why the spec only has [6 sanctioned units](https://tc39.es/proposal-unified-intl-numberformat/section6/locales-currencies-tz_proposed_out.html#table-sanctioned-simple-unit-identifiers). It's definitely a point of contention.

**Comment by TheSavior** [comment]
Hey @longlho, thanks for the response. We are definitely worried and want to do everything we can to avoid increasing the size of the binary substantially. Our current thinking is that the implementations of these APIs in Hermes would internally use the native Android APIs so that we wouldn't have to include that CLDR data. Does that seem feasible with your understanding?

**Comment by longlho** [comment]
> Hey @longlho, thanks for the response. We are definitely worried and want to do everything we can to avoid increasing the size of the binary substantially. Our current thinking is that the implementations of these APIs in Hermes would internally use the native Android APIs so that we wouldn't have to include that CLDR data. Does that seem feasible with your understanding?

Yes that's definitely feasible. Android does package `icu4j` AFAIK.

**Comment by TheSavior** [comment]
Cool, phew. :D 

This is why we are trying to figure out which APIs are the most critical to support. The Hermes implementations will require finagling the Android APIs to match the JS spec. We'd like to unblock the majority of people by prioritizing the common APIs first

**Comment by janicduplessis** [comment]
I was thinking we could strip unused locales but this is even better, I like this approach.

**Comment by longlho** [comment]
If that's your approach I think you'd basically need `DateTimeFormat` & `NumberFormat` as the bare minimums (to be on par w/ IE11 😄). `Intl.PluralRules` & `Intl.RelativeTimeFormat` are fairly small & new. `PluralRules` are mostly useful in the context of MessageFormat and rarely used raw.

**Comment by mhorowitz** [comment]
Even passing through to icu, there's still a lot of APIs.  Even if we have to do it all, it is useful to know what to prioritize.

**Comment by longlho** [comment]
The other thing to note is that `NumberFormat` in ES2015 is different from ES2017 & ES2020 😄. The tendency of ECMA-402 right now is to add more things to `options` instead of new APIs.

**Comment by janicduplessis** [comment]
Yea when talking about `XFormat` apis it would be useful to know which options are used the most (see https://github.com/facebook/hermes/issues/23#issuecomment-605340533). Or just focus on ES2015 options first.

**Comment by longlho** [comment]
If you gotta do it, I'd personally recommend stage-4 of ES2020 because right now there's no way to monkey-patch Intl APIs from ES2015 to, say ES2019, so having old support is kinda worse than no support you'd always lag behind MDN docs. Right now stage-3 polyfills are basically the full implementation.

**Comment by longlho** [comment]
Forgot another thing: have you guys run hermes through test262 yet? Intl API abstract ops rely on several building blocks within the core JS spec (ECMA262) to produce a spec-compliant output (e.g things like digit truncation, rounding & `BigDecimal`...). Therefore it's pretty important for a JS engine to pass test262 😄

**Comment by mhorowitz** [comment]
We run hermes through test262 on a continuous basis, but we skip parts we know we don't implement, like Intl.

**Comment by longlho** [comment]
Ok cool that's good to hear 👍

**Comment by mhorowitz** [comment]
@longlho I started looking into this, and I have a generic question.  ECMA 402 is written in terms of CLDR data, but icu and icu4j are written in terms of higher-level abstractions, which in turn use CLDR data underneath.  So, the APIs have strong parallels, but spec compliance is not clear.  For example, if implemented `new Intl.NumberFormat(...).format(number));` by using `com.ibm.icu.number.NumberFormatter` and it's associated classes, would this be a proper implementation?  Or am I going to run into gaps and inconsistencies?  Is `test262/test/intl402` detailed enough to provide confidence?

**Comment by longlho** [comment]
yeah test262 should be enough. Browsers all use `icu4c` under the hood so I think it should be fine for you guys (e.g V8 https://github.com/v8/v8/blob/4b9b23521e6fd42373ebbcb20ebe03bf445494f9/src/objects/js-number-format.h)

**Comment by mhorowitz** [comment]
An update on plans for Intl support.  Our primary motivation for building Intl support is to make Hermes usable as a replacement for JSC for the most apps on Android.

Currently, the android community build of JSC doesn't support all of ECMA 402.  In particular, the only properties of Intl which it supports are getCanonicalLocales, Collator, NumberFormat, DateTimeFormat, and PluralNames.   The latter was only relatively recently implemented in JSC.  In addition, DateTimeFormat.formatRange/formatRangeToParts are not implemented by JSC.  

Given the history of Intl JSC, we will plan initially to implement getCanonicalLocales, Collator, NumberFormat, and DateTimeFormat, excluding the range methods.  Later, we can add more functionality as needed.

If anybody knows of RN app which uses any other Intl APIs, please comment and tag me.  Thanks!

**Comment by gpawlik** [comment]
@mhorowitz would you be able to give us an update on the progress please?

**Comment by likern** [comment]
Is there any progress this? Is it possible to prioritize on this critical feature?

Recently highly popular react-native-reanimated library introduced new version 2 based on Turbomodules.

Now it supports on Android only Hermes engine (no JSC support yet, no recent plans). At the same time Hermes doesn't support Intl, which is highly popular in i18n (react-intl).

This leads to blocking on Intl support feature. Developers can't update their apps to Reanimated V2.

**Comment by TheSavior** [comment]
Microsoft is staffing up people to work on the Intl implementations that @mhorowitz mentioned live in https://github.com/facebook/hermes/tree/master/lib/Platform/Intl

**Comment by longlho** [comment]
FWIW we just released Intl.DateTimeFormat ES2021 polyfill so basically all the Intl APIs are polyfills except `Intl.Collator`
https://formatjs.io/docs/polyfills

**Comment by Weegle99** [comment]
> FWIW we just released Intl.DateTimeFormat ES2021 polyfill so basically all the Intl APIs are polyfills except `Intl.Collator`
> [formatjs.io/docs/polyfills](https://formatjs.io/docs/polyfills)

Hi, I tried using your polyfill to use hermes on my RN app but I cannot get it working
I keep getting Property 'Intl' doesn't exist, js engine : hermes
I don't think I missed a step though

Could you help me on the matter ? Hermes would be a huge boost on my app

**Comment by longlho** [comment]
You're looking for https://formatjs.io/docs/polyfills/intl-datetimeformat/#requirements
For Hermes it's likely you need everything in https://formatjs.io/docs/polyfills so `Intl.getCanonicalLocales`, `Intl.PluralRules`, `Intl.NumberFormat` & `Intl.DateTimeFormat`

**Comment by likern** [comment]
> > FWIW we just released Intl.DateTimeFormat ES2021 polyfill so basically all the Intl APIs are polyfills except `Intl.Collator`
> > [formatjs.io/docs/polyfills](https://formatjs.io/docs/polyfills)
> 
> Hi, I tried using your polyfill to use hermes on my RN app but I cannot get it working
> I keep getting Property 'Intl' doesn't exist, js engine : hermes
> I don't think I missed a step though
> 
> Could you help me on the matter ? Hermes would be a huge boost on my app

Please, share your experience later. Could you get it working?

**Comment by likern** [comment]
@longlho I've installed all polyfills, didn't work for me. I see this error:

```bash
[Tue Jun 23 2020 23:42:49.868]  ERROR    ReferenceError: Property 'Intl' doesn't exist, js engine: hermes
```

**Comment by longlho** [comment]
Do you have a code sample? Can you make an issue in formatjs repo?

**Comment by longlho** [comment]
Seems working for me:
![simulator](https://www.dropbox.com/s/xhvzzudm7dqbl18/Screenshot%202020-06-23%2022.36.15.png?raw=1)
![hello-world](https://www.dropbox.com/s/jhdi5ae6e6lq2fs/Screenshot%202020-06-23%2022.37.37.png?raw=1)

**Comment by likern** [comment]
> Seems working for me:
> ![simulator](https://camo.githubusercontent.com/83a9bce868884f8a6dbd9afa9ce411afe89dbb39/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f7868767a7a75646d376471626c31382f53637265656e73686f74253230323032302d30362d323325323032322e33362e31352e706e673f7261773d31)
> ![hello-world](https://camo.githubusercontent.com/de9b9db49bdab8bc06001138dad356b9edc3741b/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f6a6864693561653665366c713266732f53637265656e73686f74253230323032302d30362d323325323032322e33372e33372e706e673f7261773d31)
Could you provide repo?

**Comment by longlho** [comment]
@likern https://github.com/longlho/hermes-intl-demo

**Comment by likern** [comment]
> @likern https://github.com/longlho/hermes-intl-demo

I recreated my i18n code based on your repo https://github.com/likern/hermes-intl-demo. Looks like on expo and ejected expo it's working fine.

If I find the source of this error I let you know.

**Comment by RichardLindhout** [comment]
@likern It's working fine because you don't have Hermes enabled: https://github.com/likern/hermes-intl-demo/blob/master/android/app/build.gradle#L80.

**Comment by RichardLindhout** [comment]
For date formatting you will need at least the following now I don't get the ReferenceError: Property 'Intl' doesn't exist,  error anymore
```
import '@formatjs/intl-getcanonicallocales/polyfill'

import '@formatjs/intl-numberformat/polyfill'
import '@formatjs/intl-numberformat/locale-data/nl'

import '@formatjs/intl-datetimeformat/polyfill'
import '@formatjs/intl-datetimeformat/locale-data/nl'
```

**Comment by RichardLindhout** [comment]
I recommend my library users to use the following code (for my date picker library)

```javascript
// in your index.js file
if (!!global.HermesInternal) {
  require('@formatjs/intl-getcanonicallocales/polyfill')

  require('@formatjs/intl-pluralrules/polyfill')
  require('@formatjs/intl-pluralrules/locale-data/nl') // use your language files

  require('@formatjs/intl-relativetimeformat/polyfill')
  require('@formatjs/intl-relativetimeformat/locale-data/nl') // use your language files

  require('@formatjs/intl-listformat/polyfill')
  require('@formatjs/intl-listformat/locale-data/nl') // use your language files

  require('@formatjs/intl-displaynames/polyfill')
  require('@formatjs/intl-displaynames/locale-data/nl') // use your language files

  require('@formatjs/intl-numberformat/polyfill')
  require('@formatjs/intl-numberformat/locale-data/nl.js') // use your language files

  require('@formatjs/intl-datetimeformat/polyfill')
  require('@formatjs/intl-datetimeformat/locale-data/nl.js') // use your language files

  require('@formatjs/intl-datetimeformat/add-all-tz.js')

  require('@formatjs/intl-locale/polyfill')

  // https://formatjs.io/docs/polyfills/intl-datetimeformat/#default-timezone
  let RNLocalize = require('react-native-localize')
  if ('__setDefaultTimeZone' in Intl.DateTimeFormat) {
    Intl.DateTimeFormat.__setDefaultTimeZone(RNLocalize.getTimeZone())
  }
}

```

**Comment by mhorowitz** [comment]
@RichardLindhout Thanks, it's good to know there are workarounds while we work on Intl.

 I'm curious, what's the size impact of the JS polyfills on the app?  What if someone wants multiple languages?

**Comment by RichardLindhout** [comment]
For the multilanguage version of JSI it's 6 MB stated in the build.gradle for all locales.

I don't know what's the impact of of formatjs but I think it's a lot less since we can pick the languages needed and even if you added all languages I don't think it would be 6MB.

Maybe ask @longlho on this since he is a maintainer of formatjs.
 
It would be nice if we could specify which languages to bundle with Hermes to keep the selling point of small binaries (if it comes as close to the 6mb of the international JSI 🙄)

**Comment by longlho** [comment]
I typically recommend dynamically polyfill locale on demand. For intl-numberformat you'll likely OOM if you load everything. Note that Chrome doesn't even ship with all locales.

**Comment by longlho** [comment]
For reference this is likely the list of locales Chrome supports: https://developer.chrome.com/webstore/i18n#localeTable

**Comment by mhorowitz** [comment]
@RichardLindhout The gradle files says JSC with Intl support is 6mb larger, per architecture.  There is no such thing as an international version of JSI.

I was asking about the size of formatjs polyfills exactly because I was interested in the size impact.  Based on what @longlho says, it sounds like adding all languages will blow out your memory, so that's not a good option.  But this doesn't tell me even what a single language plus the base polyfills costs in app size.

@longlho when you say "polyfill locale on demand" do you mean add it to your application when people ask for different locales?  Since an RN app ships with its JS bundle, I'm not sure what else this might mean.

**Comment by longlho** [comment]
So to clarify I'm talking about this: https://formatjs.io/docs/polyfills/intl-relativetimeformat#dynamic-import--capability-detection

I'm not super familiar with how RN does bundling. What I meant is you can still ship w/ ALL locales, just don't load them ALL at once or it'll definitely OOM.

The core polyfill are typically small (~20 - 40KB unminified). Locale data for ALL locales are quite large
| API | Size (unminified) |
|--|--|
| Intl.getCanonicalLocales | 0 |
| Intl.Locale | 0 |
| Intl.PluralRules | 848K |
| Intl.NumberFormat | 22M |
| Intl.DateTimeFormat | 22M CLDR + 420K IANA TZ |
| Intl.RelativeTimeFormat | 3.4M |
| Intl.ListFormat | 2.2M |
| Intl.DisplayNames | 15M |

So total unminified unzipped is around **85M**. Now I did a quick gzip on `ru` which is typically the largest & most complex locale and compression shaves it down by ~83% so ballpark **14M** gzipped?

Now this is not apple to apple bc JSC is notorious for being behind on Intl APIs, so tbh I don't what's included in JSC with Intl variant

**Comment by RichardLindhout** [comment]
> @RichardLindhout The gradle files says JSC with Intl support is 6mb larger, per architecture. There is no such thing as an international version of JSI.
> 
> I was asking about the size of formatjs polyfills exactly because I was interested in the size impact. Based on what @longlho says, it sounds like adding all languages will blow out your memory, so that's not a good option. But this doesn't tell me even what a single language plus the base polyfills costs in app size.
> 
> @longlho when you say "polyfill locale on demand" do you mean add it to your application when people ask for different locales? Since an RN app ships with its JS bundle, I'm not sure what else this might mean.

I think we can't lazy load yet in RN but it would improve startup time when you have a lot of locales e.g. something like this (https://github.com/microsoft/react-native-lazy-index)

```There is no such thing as an international version of JSI.```

I did mean JSC :-D ->

```gradle
/**
 * The preferred build flavor of JavaScriptCore.
 *
 * For example, to use the international variant, you can use:
 * `def jscFlavor = 'org.webkit:android-jsc-intl:+'`
 *
 * The international variant includes ICU i18n library and necessary data
 * allowing to use e.g. `Date.toLocaleString` and `String.localeCompare` that
 * give correct results when using with locales other than en-US.  Note that
 * this variant is about 6MiB larger per architecture than default.
 */
```

**Comment by Kudo** [comment]
FWIW, current RN's jsc-android used ICU from Chromium.
Chromium maintains their locale list for android at https://chromium.googlesource.com/chromium/deps/icu.git/+/master/filters/android.json.
Different jsc-android version may use different ICU version,
here I am posting the latest Chromium list.

**Comment by longlho** [comment]
That doesn't look right. That's a language list, not locale list and they are different (e.g en vs en-GB)

**Comment by Kudo** [comment]
Oops, it was locale list before but changed to filter by language after this [CL](https://chromium.googlesource.com/chromium/deps/icu.git/+/7852ffcb31b18c49b0ec694c26095c87d046700f%5E%21/)

**Comment by longlho** [comment]
hmm I don't know enough about chromium lang filter to understand what the change entails.

**Comment by mhorowitz** [comment]
> how RN does bundling

RN apps include all the JS in the app which comes from the store.  So even if you loaded stuff as needed, your android store app with hermes + all languages would include 14mb of Intl data.  This is a significant fraction of the size of most single-ABI apps.

**Comment by longlho** [comment]
That's if you package 543 locales, which, I don't think you should or need to, 6MB worth of CLDR data would be 250 locales, which, I still don't think you should or need to support 😄 

Supporting another locale (aside from the framework & data to do so) requires operation & QA as they don't always appear the same way (think of bidi, truncation, different calendar like buddhist, different numbering system like tibet...). I haven't seen any app that supports more than a handful.

The polyfill approach has worked well for the web from what I've seen and that has a much stricter size & runtime constraint than native (e.g ECMA402 spent a lot of time discussing which `unit` are sanctioned to be part of spec bc of size).

That said, I'm not familiar with the RN ecosystem so just my 2c.

**Comment by janicduplessis** [comment]
I think the main issue with the polyfill approach and supporting a large number of locales is that currently RN doesn't have good support for code splitting and loading bundles at runtime, which I'm pretty sure is how you can scale intl polyfills on the web.

**Comment by longlho** [comment]
Yup that's right, but as i said you don't need to support 200, or even 100 locales

**Comment by RichardLindhout** [comment]
It would be really nice if we could force the locales which are needed in the `app/build.gradle` file.


It would also be nice if it would load the provided locales dynamically based on the Intl usage so it would not be in RAM.

E.g English user would never have the Dutch locales in ram.
But maybe that's standard I'm not that familiar with how engines or ram bundles work.

**Comment by mhorowitz** [comment]
Hermes is mainly a C++ library.  It doesn't control how JS is packaged or loaded.

It sounds like your suggestions are potentially relevant for React Native, if someone wanted to contribute it.

For now, our strategy for Hermes is to have the Intl APIs call into the Java android.icu APIs, so we can ship only the necessary glue between them.  I think this is more sustainable in the long run, although it might not get the functionality into the VM as quickly.

**Comment by longlho** [comment]
Yeah I agree that's the better long term solution. Does Hermes implement dynamic `import` though? That's basically the backbone of code splitting.

**Comment by mhorowitz** [comment]
> Does Hermes implement dynamic import though?

Not yet.  It's in progress at https://github.com/facebook/hermes/blob/master/doc/Features.md

**Comment by longlho** [comment]
Ok looks like those 2 combined should be future-proofing `Intl` or any future ES spec changes/polyfill interop. Polyfills are also used to correct [buggy native implementations](https://bugs.chromium.org/p/chromium/issues/detail?id=1056652&q=component%3ABlink%3EJavaScript%3EInternationalization&can=2) AND [unsupported locales](https://bugs.chromium.org/p/chromium/issues/detail?id=1056652&q=component%3ABlink%3EJavaScript%3EInternationalization&can=2) which has a fairly long release cycle as well you already mentioned.

**Comment by mhorowitz** [comment]
import is itself polyfilled, so people can still use it for those purposes.  They just can't do it dynamically.

**Comment by longlho** [comment]
@mhorowitz can you provide more details?

**Comment by mhorowitz** [comment]
@longlho About what exactly?  the import polyfill?

**Comment by longlho** [comment]
The `can't do it dynamically` part? Does that mean static `import` works but not dynamic `import`?

**Comment by mhorowitz** [comment]
We're a bit out of my domain here.  I know you can use `import` in a React Native application's JavaScript.  I believe, but I'm not sure, that this is managed by the metro bundler.  What I do know is that from the perspective of Hermes, the JavaScript is precompiled at build time to a single bytecode file, which is included in the APK.  Then at runtime, the bytecode file is mmapped into memory and processed by Hermes.  This process is designed to be very fast at runtime, which is one of Hermes's startup performance advantages.

Because all the JavaScript is baked into a single mmapped bytecode file, it's not even obvious what dynamic means.  Bytecode which hasn't been executed or data which hasn't been referenced may not even be in RAM, because pages in the mmapped file are only read on demand by the operating system.

**Comment by longlho** [comment]
By dynamic I mean the procedure described in the [spec](https://tc39.es/ecma262/#sec-hostimportmoduledynamically). Now the spec doesn't enforce `I/O` or the fact that the `Module Record` actually has to be dynamically loaded so it could be static anw. However in the interest of performance, it ideally should be dynamically loaded & evaluated from a JS engine perspective, otherwise you might get a slew of bug reported filed. In this scenario Hermes might have to re-evaluate its loading strategy. Anw sorry for digressing from the initial topic of this thread but I don't see a separate GH issue for dynamic `import` support 😢

**Comment by mhorowitz** [comment]
As I mentioned, Hermes doesn't implement `import` yet, so it's doing none of what's in the spec :-) I said we're outside of my domain because I don't personally know anything about our implementation plans or strategy for `import`.  Someone else is working on that.

My message above is attempting to describe the status quo of the current behavior of Hermes as the engine for React Native, where whatever `import` does, it's entirely out of Hermes's control.  Startup performance is one of our primary metrics, so I imagine we're doing things as lazily as possible.

**Comment by RichardLindhout** [comment]
It would be cool if Hermes could use the underlying native implementations of date so the bundle size would not grow at all for Intl support. In JavascriptCore in iOS Intl works without growing the bundle size. But I guess that won't be an options as Hermes would need to implement the Intl api for every platform.

**Comment by vonovak** [comment]
I did a small bit of researching on formatting dates using `Intl` and iOS's `NSDateFormatter` that I quickly summarized in this tweet: https://twitter.com/vonovak/status/1296454407484256257

I think there is a chance that using `Intl`, it might be impossible to get the same date formatting iOS would use in date pickers and other places in the system and other apps, so a RN app using `Int` would display dates a little different from what would be desirable if one really cares about date formatting. A workaround would be making calls across the bridge to get a date formatted (not practical) or create a `<Text />` component equivalent that would be used to render dates only (not straightforward).

Apologies if you find this a little off the topic 🙂

**Comment by longlho** [comment]
I personally find that a bit hard to believe since they're all backed by libicu

**Comment by RichardLindhout** [comment]
With Turbomodules asking the native OS what the date format would maybe be just as fast as using a polyfill. Would be interesting because that way no big bundle sizes are required

**Comment by mhorowitz** [comment]
@RichardLindhout  if you look over in the Pull Requests tab, you can see that this work is progressing.  I'm excited to see it happening!

**Comment by chrisbobbe** [comment]
> if you look over in the Pull Requests tab, you can see that this work is progressing. I'm excited to see it happening!

A few of these were recently closed! 🎉

#319 #320 #321 #322

What are the next steps? 🙂

**Comment by SrBrahma** [comment]
Is DateTimeFormat already supported?

**Comment by mhorowitz** [comment]
@SrBrahma 
> Is DateTimeFormat already supported?

None of Intl is supported yet, but it's being worked on.

**Comment by IAI95** [comment]
how to run this little code in spark ar script ?
`var now = new Date();
        now.toLocaleDateString('fa-IR');
        Diagnostics.log(now.toLocaleDateString("fa-IR"));`
is there any alternatives for that if LocaleString doesn.t work at all?

**Comment by imcyee** [comment]
To suppliment the code gist on how to polyfill, by https://github.com/facebook/hermes/issues/23#issuecomment-605333778

You can do it like this: 
import 'react-native-gesture-handler';
import { AppRegistry, LogBox } from 'react-native';
import App from './src/App';
import { name as appName } from './app.json';

```
if (global.HermesInternal) {
  // Hide timers warning (Android only)
  LogBox.ignoreLogs(['Setting a timer']);

  // Polyfills required to use Intl with Hermes engine
  require('@formatjs/intl-getcanonicallocales/polyfill');
  require('@formatjs/intl-locale/polyfill');
  require('@formatjs/intl-pluralrules/polyfill');
  require('@formatjs/intl-pluralrules/locale-data/en');
  require('@formatjs/intl-numberformat/polyfill');
  require('@formatjs/intl-numberformat/locale-data/en');
  require('@formatjs/intl-datetimeformat/polyfill');
  require('@formatjs/intl-datetimeformat/locale-data/en');
  require('@formatjs/intl-datetimeformat/add-golden-tz');
}

AppRegistry.registerComponent(appName, () => App);
```

Credit to: https://www.reddit.com/r/reactnative/comments/m3qba2/announcing_react_native_064_with_hermes_on_ios/gqsvn8i?utm_source=share&utm_medium=web2x&context=3

**Comment by RichardLindhout** [comment]
Don't forget to change timezone like described here: https://github.com/web-ridge/react-native-paper-dates#android-caveats
```
// https://formatjs.io/docs/polyfills/intl-datetimeformat/#default-timezone

  if ('__setDefaultTimeZone' in Intl.DateTimeFormat) {
  
    // If you are using react-native-cli
    let RNLocalize = require('react-native-localize');
    Intl.DateTimeFormat.__setDefaultTimeZone(RNLocalize.getTimeZone());
    
    //  Are you using Expo, use this instead of previous 2 lines
    //  Intl.DateTimeFormat.__setDefaultTimeZone(
    //    require("expo-localization").timezone
    //  );
  }
```

I don't get why Github hides comments in the middle which were an important part of this conversation..

**Comment by andreialecu** [comment]
We've been using the polyfills from `@formatjs`, similar to https://github.com/facebook/hermes/issues/23#issuecomment-799253902.

While investigating some severe performance issues I noticed that the polyfills are extremely slow.

Sample benchmark using `luxon` and `@formatjs` on Android via Hermes:

```ts
  for (let y = 0; y < 5; y++) {
    const start = Date.now();
    for (let i = 0; i < 100; i++) {
      DateTime.local().toLocaleString();
    }
    console.log(Date.now() - start);
  }
```

Output:

```
LOG  423
LOG  450
LOG  575
LOG  450
LOG  421
```

Notice how formatting even a single date takes about **5ms(!)**. If you need to format multiple dates, and don't use any sort of caching, this will severely impact your application's performance.

Native `Intl` support in Hermes would probably circumvent this issue.

**Comment by longlho** [comment]
Are you sure it's not `luxon`'s issue: https://codesandbox.io/s/jovial-mestorf-s11tu?file=/src/index.js logs out like 65 across multiple browsers for me (so 0.65ms per `Date`).

**Comment by andreialecu** [comment]
@longlho I used a profiler to check and most of the time is spent in various functions in `formatjs`.

I ran the same benchmark using JSC instead of Hermes on Android and the times are 150-170ms per 100 runs. About 3 times faster than Hermes.

So it seems to be a Hermes performance issue as well, probably same root cause as #471 

I put the repro on GitHub:
https://github.com/andreialecu/hermes-repro/tree/jsc-intl
https://github.com/andreialecu/hermes-repro/tree/hermes-intl

The same benchmark on an iOS emulator (iOS has native Intl so it does not need a polyfill) using JavaScriptCore logs:

```
 LOG  8
 LOG  0
 LOG  1
 LOG  1
 LOG  0
```

Faster than desktop browsers.

**Comment by rosskhanas** [comment]
I can confirm that there is an issue with a https://formatjs.io/docs/polyfills/intl-datetimeformat polyfill performance.

**Comment by longlho** [comment]
Yeah polyfills will never be as fast as native unfortunately. If you cache the constructor though you'd virtually have no perf issue.

**Comment by andreialecu** [comment]
@longlho what do you mean by caching the constructor? 

In our app, we need to display tens or hundreds of formatted dates, most will be different. It's the `toLocaleString()` call that is the bottleneck. Not constructing luxon dates. Luxon is actually absolutely negligent.

In a real app the `.toLocaleString()` are basically spread across various components (no one thought they'd be a performance issue). So that makes the performance problem substantially worse. There's a 5ms lock in various render functions, etc.

Now with RN 0.64, Hermes is available for iOS, so the lack of Intl will start becoming a problem on iOS as well.

**Comment by longlho** [comment]
toLocaleDateString is just a wrapper on top of Intl.DateTimeFormat and it initializes a new object every call. You should cache Intl.DateTimeFormat and reuse it to format your dates

**Comment by andreialecu** [comment]
@longlho I assume you meant this:

```ts
  start = Date.now();
  const dtf = new Intl.DateTimeFormat('en');
  for (let i = 0; i < 100; i++) {
    dtf.format(new Date());
  }
  results.dateTimeFormat.push(Date.now() - start);
```

Makes no difference. Exact same times logged. ~500ms per 100 runs on Hermes, ~150 per 100 on JSC.

**Comment by longlho** [comment]
hmm weird, it's like at least several magnitude faster for me.

**Comment by andreialecu** [comment]
It's only about 2x faster in a browser: 
https://codesandbox.io/s/determined-nash-j9wug?file=/src/index.js
versus
https://codesandbox.io/s/jovial-mestorf-s11tu?file=/src/index.js

However, I don't see the same 2x improvement with react-native. There's no difference there.

You can try it yourself, I pushed a commit with this new benchmark:
https://github.com/andreialecu/hermes-repro/commit/4910fd72a0c9dcf9b895af3d5c48f6875269f121

Output on Android JSC:
```
{"dateTimeFormat": [919, 175, 162, 158, 147, 159, 154, 149, 156, 148], "local": [11, 0, 0, 1, 0, 0, 0, 0, 0, 0], "setZone": [1381, 184, 172, 167, 155, 183, 150, 152, 151, 154], "toLocaleString": [517, 175, 173, 165, 158, 157, 147, 148, 152, 147]}
```

Output on Hermes:
{"dateTimeFormat": [393, 373, 369, 343, 341, 349, 372, 364, 347, 343], "local": [5, 1, 0, 0, 1, 1, 0, 1, 0, 1], "setZone": [401, 376, 376, 355, 355, 361, 389, 379, 364, 352], "toLocaleString": [375, 392, 368, 368, 342, 392, 367, 379, 348, 342]}
(a bit faster than 500ms on this run, I guess an emulator variation)

**Comment by andreialecu** [comment]
Here's the output of the benchmark running on a mid-level Android phone from 2020:

```
{
"dateTimeFormat": [2006, 2016, 2005, 2009, 2004, 2060, 2008, 2012, 2020, 2017], 
"local": [28, 3, 3, 2, 4, 2, 3, 3, 3, 3], 
"setZone": [2214, 2077, 2073, 2067, 2075, 2129, 2080, 2078, 2075, 2077], 
"toLocaleString": [2042, 2035, 2003, 2019, 2024, 2025, 2029, 2035, 2011, 2013]}
```

(A lot slower than running it native arm64 Android Emulator on the M1 cpu, as expected)

That's **20ms** per date. Only 50 can be processed per second. But in reality much fewer because they're usually not batched.

A bit of a deal-breaker for apps that need to show list of events, calendars, and other similar things 🙁 .

**Comment by Huxpro** [comment]
Hey folks, thanks for all the patience. Here is a spoiler: We plan to release Intl support in Hermes on Android with the upcoming RN 0.65. It would be enabled and available by default since the binary size increase is fairly acceptable (57-62K per ABI, comparing to [JSC's 6MiB difference](https://github.com/react-native-community/jsc-android-buildscripts). People are encouraged to try and give us feedbacks during the RN0.65 rc periods. Stay tuned!

**Comment by andreialecu** [comment]
@Huxpro thanks for the update. Will support for DateTimeFormat and RelativeTimeFormat be included, or just numerics?

**Comment by Huxpro** [comment]
@andreialecu 

> DateTimeFormat and RelativeTimeFormat be included, or just numerics?

DateTimeFormat for sure. RelativeTimeFormat, afaik, is still at Stage-3.

In general, **we supports almost everything from ECMA-402 up to ES2020**. There are a few of known incompetences and inconsistencies (imposed by the limitations and differences between older and newer Android SDK versions). 

We will come out detailed documents about our support and we may need helps from the community to come out either polyfills or maybe they are just too edge-casy and no one care 😛 .

**Comment by seamive** [comment]
Has it been added?

**Comment by mganandraj** [comment]
Here is a short writeup on the Intl implementation on Android: https://github.com/facebook/hermes/raw/master/lib/Platform/Intl/java/com/facebook/hermes/intl/HermesIntlAPIsImplementationNotes.docx

**Comment by mrousavy** [comment]
Just to confirm, Intl support will be Android only right? Haven't read any mentions about iOS in that paper 🤔

**Comment by andreialecu** [comment]
An update for people relying on the formatjs polyfill. I was able to find a big bottleneck in the code:
https://github.com/formatjs/formatjs/pull/2827

Here's the new benchmark result using Hermes on the same Android device that was taking 20ms per format before (https://github.com/facebook/hermes/issues/23#issuecomment-816006209):

```
is it HERMES? function foo(a0, a1, a2) { [bytecode] }
result true
{
"dateTimeFormat": [34, 35, 35, 35, 35, 35, 34, 35, 34, 35], 
"local": [28, 3, 3, 2, 3, 3, 2, 3, 2, 2], 
"setZone": [143, 77, 77, 79, 78, 78, 79, 78, 79, 78], 
"toLocaleString": [48, 39, 38, 39, 39, 38, 38, 39, 39, 39]
}
```

100-200x(!) faster. Only 0.3ms per format now. This should make the polyfills usable and support for native Intl no longer a blocking issue.

/cc @rosskhanas

**Comment by msvargas** [comment]
> An update for people relying on the formatjs polyfill. I was able to find a big bottleneck in the code:
> [formatjs/formatjs#2827](https://github.com/formatjs/formatjs/pull/2827)
> 
> Here's the new benchmark result using Hermes on the same Android device that was taking 20ms per format before ([#23 (comment)](https://github.com/facebook/hermes/issues/23#issuecomment-816006209)):
> 
> ```
> is it HERMES? function foo(a0, a1, a2) { [bytecode] }
> result true
> {
> "dateTimeFormat": [34, 35, 35, 35, 35, 35, 34, 35, 34, 35], 
> "local": [28, 3, 3, 2, 3, 3, 2, 3, 2, 2], 
> "setZone": [143, 77, 77, 79, 78, 78, 79, 78, 79, 78], 
> "toLocaleString": [48, 39, 38, 39, 39, 38, 38, 39, 39, 39]
> }
> ```
> 
> 50-100x(!) faster. Only 0.3ms per format now. This should make the polyfills usable and support for native Intl no longer a blocking issue.
> 
> /cc @rosskhanas

Great!, good job, How can I add those changes to my project? any patch file? thanks!

**Comment by sflahave** [comment]
Sorry if this is off-topic, but I'm wondering what the implications are for unit testing. Suppose you have a `time.js` util file that has a bunch of functions for date/time manipulation and formatting, using something like DayJS or some other library that assumes support for Intl (or at least a polyfill).  Suppose you use Jest to unit test these functions. 
Suppose for some reason you have enabled Hermes on iOS but not on Android. What can you conclude about the success or failure of these unit tests? In other words, can you have _any_ confidence in _any_ unit tests if the test runner doesn't use the same JS engine that the app would?

**Comment by mhorowitz** [comment]
You could mock the Intl functions.  This would isolate the test/failure of your utilities from the specific Intl implementation you're working with.

**Comment by RichardLindhout** [comment]
> Sorry if this is off-topic, but I'm wondering what the implications are for unit testing. Suppose you have a `time.js` util file that has a bunch of functions for date/time manipulation and formatting, using something like DayJS or some other library that assumes support for Intl (or at least a polyfill). Suppose you use Jest to unit test these functions.
> Suppose for some reason you have enabled Hermes on iOS but not on Android. What can you conclude about the success or failure of these unit tests? In other words, can you have _any_ confidence in _any_ unit tests if the test runner doesn't use the same JS engine that the app would? What engine does the test runner use?

This is always a problem anyway if you really want to test the final output on a real device / real browser you should use integration tests like Detox https://github.com/wix/Detox

**Comment by wilau2** [comment]
> Hey folks, thanks for all the patience.
> 
> Here is a spoiler: **We plan to release Intl support in Hermes on Android with the upcoming RN 0.65.** It would be enabled and available by default since the binary size increase is fairly acceptable (57-62K per ABI, comparing to [JSC's 6MiB difference](https://github.com/react-native-community/jsc-android-buildscripts)). People are encouraged to try and give us feedbacks during the RN 0.65 rc periods. We'll post more details later. Stay tuned!

Hey just wondering what is the plan for application out there already using `Hermes & fbt` as I thought that was the direction taken to have intl with Hermes. Will intl in hermes be optional with `0.65` ?

**Comment by Huxpro** [comment]
@wilau2 

> Will intl in hermes be optional with 0.65 ?

Was your main concern the presence of `Intl` API or the extra binary size? The plan is that we would only ship `Intl`-included pre-compiled binary to the `hermes-engine` npm package and one may _hide_ the `Intl` APIs via `RuntimeConfig` but that won't reduce the code size since `Intl` is already compiled-in (It will remove some initialization overhead though). To get a `Intl`-excluded binary you will have to compile Hermes from the source with the CMake flag `HERMES_ENABLE_INTL=FALSE`.

**Comment by mhorowitz** [comment]
To add to @Huxpro's answer:

If you're using fbt today, it should keep working.  If that is not the case, please open a new issue for that so we can investigate.

We ran a poll on Twitter a while back to see how the community felt about including Intl in our prebuilt npm by default, and the results were strongly in favor: https://twitter.com/tmikov/status/1336442786694893568.  The tweet links to https://github.com/facebook/hermes/pull/426 with more detailed numbers.

**Comment by wilau2** [comment]
My concern is more about knowing what is recommended. 
- we used react-intl before Hermes
- migrated to fbt to use Hermes
- now that Hermes supports intl I am wondering what the recommended approach is.
- What will facebook use in its app going forward?
- Should we do some performance benchmark of ftb vs intl in js?

**Comment by mhorowitz** [comment]
I expect FB will continue using fbt, as we depend on the text translation features as well, and Intl doesn't do that.

I don't think we have a single recommendation, or are likely to make one.  JS supports a lot of stuff, and people should use whatever works best for them.  If you are concerned about performance, then you should probably benchmark the two approaches.  I'm sure others would be interested in seeing your results!

**Comment by TheSavior** [comment]
To add to what @mhorowitz said, FB has a lot of constraints that other people don't have. For example, most of our UI is powered by the server. We get text strings back from the server on how to display internationalized stuff so that all of our clients can be consistent. We do very little local I18n processing, which is why we are able to rely on fbt. If you don't have those same constraints, something like the Intl implementation in Hermes might make more sense for you than it does for Facebook.

**Comment by Jpunt** [comment]
FYI: In order to get [date-fns-tz](https://www.npmjs.com/package/date-fns-tz) to work with Hermes on Android, we had to do the following:

```
npm i intl @formatjs/intl-locale @formatjs/intl-datetimeformat
```

In index.js:

```
if (global.HermesInternal) {
  if (typeof Intl === "undefined") {
    require("intl")
    require("intl/locale-data/jsonp/nl") // may be different for you
  }
  require("@formatjs/intl-locale/polyfill")
  require("@formatjs/intl-datetimeformat/polyfill")
  require("@formatjs/intl-datetimeformat/locale-data/nl") // may be different for you
  require("@formatjs/intl-datetimeformat/add-all-tz")
}
```

**Comment by longlho** [comment]
@Jpunt you should change
```js
 if (typeof Intl === "undefined") {
    require("intl")
    require("intl/locale-data/jsonp/nl") // may be different for you
  }
```
to just
```
require('@formatjs/intl-getcanonicallocales');
```

**Comment by andreialecu** [comment]
>   require("@formatjs/intl-datetimeformat/polyfill")
>   require("@formatjs/intl-datetimeformat/locale-data/nl") // may be different for you
>   require("@formatjs/intl-datetimeformat/add-all-tz")
> }
> ```

@Jpunt be careful with the formatjs datetimeformat polyfill as it is extremely slow and the screens you're showing formatted dates on will likely be extremely laggy.

There is some discussion further above in this thread on this subject. I'm currently using this patch to make it bearable to use: https://gist.github.com/andreialecu/2388730f9bde365d934cb343ecc38102

**Comment by longlho** [comment]
that will be fixed once https://github.com/tc39/ecma402/pull/569 is merged. We've had consensus from the committee.

**Comment by a-leblond** [comment]
> that will be fixed once [tc39/ecma402#569](https://github.com/tc39/ecma402/pull/569) is merged. We've had consensus from the committee.

Hi, will this be delivered as a fix in RN 0.65.x ?

**Comment by Bardiamist** [comment]
By some reason `''.toLocaleLowerCase()` returns `뜂�띑┰ᰀ궜 瀁┰ ࿑鯮�띑┰᱀궜࿑鯮�뜂₠럌랇�뜂謂띡�` on some Android devices (React native 0.65.1). Ok, I'll use `toLowerCase`.

**Comment by idutka** [comment]
Is there a chance that Hermes will support Intl on iOS?

**Comment by matinzd** [comment]
I had to require intl package with locales to make it work with hermes. 

```if (typeof Intl === 'undefined') {
  require('intl')
  require('intl/locale-data/jsonp/en')
  require('intl/locale-data/jsonp/sv')
  require('intl/locale-data/jsonp/es')
}```

**Comment by Huxpro** [comment]
Hey @idutka 

> Is there a chance that Hermes will support Intl on iOS?

It's definitely on our radar and we are looking into it! We don't have a timeline yet but we will keep everyone here posted when we have any update ;)

**Comment by SudoPlz** [comment]
Do we need to close that since Intl is now part of hermes 0.8.1 on react-native?

**Comment by neildhar** [comment]
Update: work on iOS Intl support is currently underway with @silva-nick and @SarahCoded implementing it as part of an MLH Open Source Fellowship!

**Comment by cristianoccazinsp** [comment]
> require('@formatjs/intl-pluralrules/locale-data/en').default;

How do you install these? Do you really need to install a dependency for every single locale you want to support? With JSC, you can pass pretty much any locale and it works, which is great to support currency formatting on the device's locale without effort.

**Comment by cristianoccazinsp** [comment]
@neildhar looking forward to that. The polyfills are awful.

**Comment by andreialecu** [comment]
Hermes 0.10.0 finally fixes all outstanding Android Intl issues. (ref: https://github.com/facebook/hermes/commit/9238467f1cceead5426bca8ef5c794137d504429)

I've been able to use yarn `resolutions` to force Hermes to version 0.10.0 even on RN 0.66 via:

```
  "resolutions": {
    "hermes-engine": "0.10.0"
  }
```

This allowed us to finally drop the polyfills and there's an **amazing** performance boost on Android as a result in our app. 🎊

**Comment by mattijsf** [comment]
> Hermes 0.10.0 finally fixes all outstanding Android Intl issues. (ref: [9238467](https://github.com/facebook/hermes/commit/9238467f1cceead5426bca8ef5c794137d504429))
> 
> I've been able to use yarn `resolutions` to force Hermes to version 0.10.0 even on RN 0.66 via:
> 
> ```
>   "resolutions": {
>     "hermes-engine": "0.10.0"
>   }
> ```
> 
> This allowed us to finally drop the polyfills and there's an **amazing** performance boost on Android as a result in our app. 🎊

Until you need timezone conversion support though 👀

**Comment by andreialecu** [comment]
> Until you need timezone conversion support though 👀

The commit I linked specifically fixes the bug in timezone conversion support that was present in 0.9.0.

**Comment by mattijsf** [comment]
> > Until you need timezone conversion support though 👀
> 
> The commit I linked specifically fixes the bug in timezone conversion support that was present in 0.9.0.

Thanks I was too quick with my comment sorry. Unfortunately polyfills are still necessary on iOS to be able to do the timezone conversion (when using Hermes). I read above that this is being worked on 🤞

**Comment by billnbell** [comment]
Question with hermes in v0.67.1 react-native enabled do I need this?

```
if (global.HermesInternal) {
  if (Platform.OS === 'ios') {
    // Polyfills required to use Intl with Hermes engine
    require('@formatjs/intl-getcanonicallocales/polyfill').default;
    require('@formatjs/intl-locale/polyfill').default;
    require('@formatjs/intl-pluralrules/polyfill').default;
    require('@formatjs/intl-pluralrules/locale-data/en').default;
    require('@formatjs/intl-numberformat/polyfill').default;
    require('@formatjs/intl-numberformat/locale-data/en').default;
    require('@formatjs/intl-datetimeformat/polyfill').default;
    require('@formatjs/intl-datetimeformat/locale-data/en').default;
    require('@formatjs/intl-datetimeformat/add-all-tz').default;
  } else {
    require('@formatjs/intl-getcanonicallocales/polyfill');
    require('@formatjs/intl-locale/polyfill');
    require('@formatjs/intl-datetimeformat/polyfill');
    require('@formatjs/intl-datetimeformat/locale-data/en');
    require('@formatjs/intl-datetimeformat/add-all-tz');
  }
}
```

**Comment by cristianoccazinsp** [comment]
What is the progress on this? INTL support is the only release blocker for us to move to Hermes on iOS.  The polyfills are way too slow and incorrect.

**Comment by tranducduong1994** [comment]
Any update on IOS Intl support?

**Comment by pke** [comment]
@andreialecu 

> Hermes 0.10.0 finally fixes all outstanding Android Intl issues. (ref: [9238467](https://github.com/facebook/hermes/commit/9238467f1cceead5426bca8ef5c794137d504429))
> 
> I've been able to use yarn `resolutions` to force Hermes to version 0.10.0 even on RN 0.66 via:
> 
> ```
>   "resolutions": {
>     "hermes-engine": "0.10.0"
>   }
> ```

Does this make this obsolete?

``` 
 if (typeof Intl === "undefined") {
    require("intl")
    require("intl/locale-data/jsonp/nl") // may be different for you
  }
```

**Comment by andreialecu** [comment]
@pke: Most polyfills should no longer be required with the updated Hermes. So yes, those two should be obsolete.

You should check however if things work properly after removing them. I think `RelativeTimeFormat` was not implemented by Hermes yet, for example.

**Comment by pke** [comment]
@andreialecu how do I know which Hermes engine is currently bundled with my RN project? 
And is there a List of not yet implemented functions like RelativeTimeFormat (which we do not use atm)?

**Comment by mrousavy** [comment]
> how do I know which Hermes engine is currently bundled with my RN project?
You can check `yarn.lock` and search for `hermes-engine`.

**Comment by francois-pasquier** [comment]
is iOS still needing polyfills with latest hermes?

**Comment by nandorojo** [comment]
> > how do I know which Hermes engine is currently bundled with my RN project?
> 
> You can check `yarn.lock` and search for `hermes-engine`.

`yarn why hermes-engine`

**Comment by neildhar** [comment]
> is iOS still needing polyfills with latest hermes?

As of now, yes, although we are actively working on Intl support. Support for the upper/lowercase conversion functions, getCanonicalLocales, and Collator has already been merged in trunk.

**Comment by zhiqingchen** [comment]
Same issue, Using
`new Date().toLocaleTimeString()` 

will cause

Error: RangeError: java.lang.ClassNotFoundException: Didn't find class "com.facebook.hermes.intl.DateTimeFormat"

**Comment by gavinaboulhosn** [comment]
> Same issue, Using `new Date().toLocaleTimeString()`
> 
> will cause
> 
> Error: RangeError: java.lang.ClassNotFoundException: Didn't find class "com.facebook.hermes.intl.DateTimeFormat"

Make sure you're including the Hermes android artifacts on the classpath

**Comment by billnbell** [comment]
> Question with hermes in v0.67.1 react-native enabled do I need this?
> 
> ```
> if (global.HermesInternal) {
>   if (Platform.OS === 'ios') {
>     // Polyfills required to use Intl with Hermes engine
>     require('@formatjs/intl-getcanonicallocales/polyfill').default;
>     require('@formatjs/intl-locale/polyfill').default;
>     require('@formatjs/intl-pluralrules/polyfill').default;
>     require('@formatjs/intl-pluralrules/locale-data/en').default;
>     require('@formatjs/intl-numberformat/polyfill').default;
>     require('@formatjs/intl-numberformat/locale-data/en').default;
>     require('@formatjs/intl-datetimeformat/polyfill').default;
>     require('@formatjs/intl-datetimeformat/locale-data/en').default;
>     require('@formatjs/intl-datetimeformat/add-all-tz').default;
>   } else {
>     require('@formatjs/intl-getcanonicallocales/polyfill');
>     require('@formatjs/intl-locale/polyfill');
>     require('@formatjs/intl-datetimeformat/polyfill');
>     require('@formatjs/intl-datetimeformat/locale-data/en');
>     require('@formatjs/intl-datetimeformat/add-all-tz');
>   }
> }
> ```

OK, with 0.68.1 which of these can I drop?

**Comment by mattijsf** [comment]
@billnbell For me it's also unclear to what extent Intl is supported on both platforms. I would love to see some official support matrix of Hermes features for each platform per Hermes version.

**Comment by Jackman3005** [comment]
@mattijsf  They have a [Intl APIs page](https://github.com/facebook/hermes/blob/main/doc/IntlAPIs.md) that covers what is supported/not yet supported and limitations of different versions of Android. But I believe __there is still no iOS support for `Intl` in hermes__. 

I'm using a [time zone library](https://github.com/vvo/tzdb) for displaying a list of time zones and it has a reasonable usage of `Intl` that is currently causing zero timezones to be retrieved in iOS due to the lack of support for `Intl`. Looks like I'll need to polyfill for iOS until this is supported with hermes. I will try something along the lines of what @billnbell has done.

Thanks for all the hard work on Hermes so far, you folks rock. I'm looking forward to when (soon?) `Intl` support is solid for both platforms.

**Comment by zbraniecki** [comment]
hi all. I'm one of the founders of the [ICU4X project](https://github.com/unicode-org/icu4x) which is a Rust implementation of ICU targeting many programming environments and aiming for the ECMA-402 scope.

One of the interesting features of ICU4X is that one can pull small modules that provide necessary components without paying for the whole monolithic codebase, and compile those components to WASM or JS.

We are planning to release 1.0 quite soon and I'm wondering if Hermes would be interested in evaluating ICU4X for your needs.

**Comment by jkester1986** [comment]
We just turned off Hermes in our iOS app after recently turning it on due to discovering lack of intl support for dates/times. Do we have any ETA on when this will be supported on the iOS side?

**Comment by jpporto** [comment]
Hi @jkester1986,

The Hermes Intl support (for date and time) is already implemented, but some wiring is needed on RN to surface it. We expect the next RN version (i.e., 0.70) to have that plumbing done, and full feature parity with Hermes on Android for RN 0.71

**Comment by jpporto** [comment]
Hi @billnbell,

Hermes' iOS Intl support is defined [here](https://github.com/facebook/hermes/blob/main/lib/Platform/Intl/PlatformIntlApple.mm). It seems currently, Number, DateTime, and Collator are implemented. Other missing features are being actively implemented.

**Comment by datanaman** [comment]
Hi Team , we are still facing the same issue only on IOS after enabling Hermes : 
Issue : 
ReferenceError: Property 'Intl' doesn't exist, js engine: hermes
 ERROR  Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication). A frequent cause of the error is that the application entry file path is incorrect.

 Version : "react-native": "0.68.2"

<img width="998" alt="image" src="https://user-images.githubusercontent.com/34430404/175342057-51fbfcc7-f337-4ff9-87d6-9cadf4a51092.png">


Tried the below workaround in out index.ts - but doesn't seem to work.

import { AppRegistry, Platform } from 'react-native';
// eslint-disable-next-line no-undef
const usingHermes = typeof HermesInternal === 'object' && HermesInternal !== null;
if (usingHermes && Platform.OS === 'ios') {
  // Polyfills required to use Intl with Hermes engine
  require('@formatjs/intl-getcanonicallocales/polyfill');
  require('@formatjs/intl-locale/polyfill');
  require('@formatjs/intl-pluralrules/polyfill');
  require('@formatjs/intl-pluralrules/locale-data/en');
  require('@formatjs/intl-numberformat/polyfill');
  require('@formatjs/intl-numberformat/locale-data/en');
  require('@formatjs/intl-datetimeformat/polyfill');
  require('@formatjs/intl-datetimeformat/locale-data/en');
  require('@formatjs/intl-datetimeformat/add-golden-tz');

  //require('intl'); // import intl object
}

import App from './src/App';
import { name as appName } from './app.json';

AppRegistry.registerComponent(appName, () => App);

**Comment by billnbell** [comment]
You are not putting it in right spot. Put in index.js:

```
/* eslint-disable no-unused-expressions */
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable global-require */
import React from 'react';
import { AppRegistry, Platform } from 'react-native';

import { expo } from './app.json';
import App from './src/App';
import AppFake from './src/AppFake';
import { configureCalendarLocale } from './src/components/member/explore/property-screen/dateHelpers';
import { setBackgroundMessageHandler } from './src/utils/notificationHandler';

const { name: appName } = expo;

if (global.HermesInternal) {
  if (Platform.OS === 'ios') {
    // Polyfills required to use Intl with Hermes engine
    require('@formatjs/intl-getcanonicallocales/polyfill').default;
    require('@formatjs/intl-locale/polyfill').default;
    require('@formatjs/intl-pluralrules/polyfill').default;
    require('@formatjs/intl-pluralrules/locale-data/en').default;
    require('@formatjs/intl-numberformat/polyfill').default;
    require('@formatjs/intl-numberformat/locale-data/en').default;
    require('@formatjs/intl-datetimeformat/polyfill').default;
    require('@formatjs/intl-datetimeformat/locale-data/en').default;
    require('@formatjs/intl-datetimeformat/add-all-tz').default;
  } else {
    require('@formatjs/intl-getcanonicallocales/polyfill');
    require('@formatjs/intl-locale/polyfill');
    require('@formatjs/intl-datetimeformat/polyfill');
    require('@formatjs/intl-datetimeformat/locale-data/en');
    require('@formatjs/intl-datetimeformat/add-all-tz');
  }
}
configureCalendarLocale();
try {
  setBackgroundMessageHandler();
} catch (err) {
  console.log('index.js setBackgroundMessageHandler', err);
}
const HeadlessCheck = (props) => {
  // eslint-disable-next-line react/prop-types
  const { isHeadless } = props;
  if (Platform.OS === 'ios' && isHeadless) {
    return <AppFake />;
  }
  // eslint-disable-next-line react/jsx-props-no-spreading
  return <App {...props} />;
};

AppRegistry.registerComponent(appName, () => HeadlessCheck);

```

**Comment by datanaman** [comment]
@billnbell i had it at index.ts/index.js , but still doesn' t work.

**Comment by nialldbarber** [comment]
Currently on Hermes `0.9.0` with React Native `0.67.4`.

For those using the `@formatjs polyfill`, can you confirm if you need to add this in your Hermes check as detailed in [this answer](https://github.com/facebook/hermes/issues/23#issuecomment-675098624)? 

```jsx
// https://formatjs.io/docs/polyfills/intl-datetimeformat/#default-timezone
let RNLocalize = require('react-native-localize')
if ('__setDefaultTimeZone' in Intl.DateTimeFormat) {
  Intl.DateTimeFormat.__setDefaultTimeZone(RNLocalize.getTimeZone())
}
```

It looks like (from the [link posted](https://github.com/web-ridge/react-native-paper-dates#android-caveats) in a [follow up](https://github.com/facebook/hermes/issues/23#issuecomment-802339643)), that it is only for Android, which has stable support for `Intl`. But just want to make sure that you don't need it for iOS?

**Comment by pedpess** [comment]
@nialldbarber these changes are for iOS only as `Intl` support for Hermes is just available for Android. However, on RN `v0.70.0` the missing support for iOS landed, so updating straight to that the polyfills won't be needed anymore.

**Comment by jonathanj** [comment]
Maybe worth pointing out that iOS's implementation is [still missing](https://github.com/facebook/hermes/blob/163f70598f88a11c0fdd171de288341814217584/lib/Platform/Intl/PlatformIntlApple.mm#L2519-L2522) support for `NumberFormat.formatToParts`.

**Comment by tamlyn** [comment]
Some other NumberFormat features are missing too:

https://github.com/facebook/hermes/blob/163f70598f88a11c0fdd171de288341814217584/lib/Platform/Intl/PlatformIntlApple.mm#L2444-L2449

**Comment by billnbell** [comment]
PK so we need a comprehensive list. What is the delta between Android and IOS?

**Comment by MaxMaesLW** [comment]
I'm on RN0.70 not all formats are supported for my locale (cccc returns null). I added the formatjs polyfills as described above. But when I format a (Luxon) DateTime object at first date of month, at exactly 00:00 with a positive timezone offset (for example ISO 2022-11-09T00:00:00.000+01:00), it gives me the previous month (so 'october' for the example ISO string).

Adding the TZ offset to the hours value of the datetime object seems to 'fix' it, but this shouldn't be an issue as the month property returns the proper numeric month value.

Disabling the polyfills makes it return the proper values though, so I'm not sure if this is an hermes issue or issue at formatjs...

Anyone got some thoughts on this?

**Comment by dlebedynskyi** [comment]
Is there a plan to support `dateStyle` and `timeStyle` on Android? 

It is on iOS, but on Android poly fill is still required, if you try to those options

**Comment by mganandraj** [comment]
> 

It should be a fairly straightforward to add the support for dateStyle/timeStyle form Android platform 24+ .. 

Could you file an issue if one doesn't exist? A PR would be welcome as well  !

**Comment by dlebedynskyi** [comment]
@mganandraj created https://github.com/facebook/hermes/issues/926.

**Comment by joluet** [comment]
> Maybe worth pointing out that iOS's implementation is [still missing](https://github.com/facebook/hermes/blob/163f70598f88a11c0fdd171de288341814217584/lib/Platform/Intl/PlatformIntlApple.mm#L2519-L2522) support for `NumberFormat.formatToParts`.

Is it planned to add this any time soon?

**Comment by arnolicious** [comment]
Can anyone else confirm that only on iOS, `DateTimeFormat` seems to not support any TimeZone Names starting with `Etc`, such as `Etc/GMT`, `Etc/UTC` `Etc/Greenwich` ?

All other TZs seem to work (`Europe/...`, `Atlantic/...` etc)

**Comment by billnbell** [comment]
Any new news?

**Comment by tmikov** [comment]
Closing this, since Intl support has been added to Hermes. Please file separate issues for individual problems (of which there are many, since the implementation is not complete).
